filter-accela-hamilton/accela_hamilton.cpp

```diff
@@ -0,0 +1,115 @@
+/* Copyright (c) 2012-2015 Stanislaw Halik
+ * Copyright (c) 2023-2024 Michael Welter
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ */
+#include "accela_hamilton.h"
+#include "compat/math.hpp"
+#include "api/plugin-api.hpp"
+#include "opentrack/defs.hpp"
+
+#include <algorithm>
+#include <QDebug>
+#include <QMutexLocker>
+
+#include "compat/math-imports.hpp"
+#include "compat/time.hpp"
+
+
+accela_hamilton::accela_hamilton()
+{
+    s.make_splines(spline_rot, spline_pos);
+}
+
+
+void accela_hamilton::filter(const double* input, double *output)
+{
+    constexpr double EPSILON = 1e-30;
+
+    const QQuaternion current_rot = QQuaternion::fromEulerAngles(input[Pitch], input[Yaw], input[Roll]);
+    const QVector3D current_pos(input[TX], input[TY], input[TZ]);
+
+    if (unlikely(first_run))
+    {
+        first_run = false;
+        last_rotation = current_rot;
+        last_position = current_pos;
+        t.start();
+#if defined DEBUG_ACCELA
+        debug_max = 0;
+        debug_timer.start();
+#endif
+        return;
+    }
+
+    const double pos_thres{s.pos_smoothing};
+    const double pos_dz{ s.pos_deadzone};
+
+    const double dt = t.elapsed_seconds();
+    t.start();
+
+    // Position
+    {
+        const QVector3D delta = current_pos - last_position;
+        const float delta_len = delta.length();
+        QVector3D delta_normed = delta_len>0.F ? delta/delta_len : QVector3D(); // Zero vector when length was zero.
+        const double gain = dt*spline_pos.get_value_no_save(std::max(0., delta_len-pos_dz) / pos_thres);
+        const QVector3D output_pos = last_position + gain * delta_normed;
+        output[TX] = output_pos.x();
+        output[TY] = output_pos.y();
+        output[TZ] = output_pos.z();
+        last_position = output_pos;
+    }
+
+    // Zoom smoothing:
+    const double max_zoomed_smoothing {s.max_zoomed_smoothing};		
+    const double max_z {s.max_z};
+
+    const double zoomed_smoothing = output[TZ]>0. ? 0. : std::min(1., -output[TZ] / (max_z + EPSILON))*max_zoomed_smoothing;
```

Comment:
Can you make it look less scary?

Reply:
Like it is now?

Reply:
Still don't like it, but it wouldn't be productive for me to keep nitpicking it. You know what you're doing.

Reply:
:-O